In [58]:
%load_ext nb_black

The nb_black extension is already loaded. To reload it, use:
  %reload_ext nb_black


<IPython.core.display.Javascript object>

In [66]:
from torchsummary import summary
import torch.optim as optim
import torch.nn.functional as F
import models
import utils
import torch

torch.manual_seed(1)

<IPython.core.display.Javascript object>

In [83]:
from importlib import reload

reload(models)
reload(utils)
reload(config)

<module 'config' from '/mnt/e/TSAI/EVA8/eva8-session7/config.py'>

<IPython.core.display.Javascript object>

# Model 1

## Target :- 

- Take the model from last week, and observe the maximum validation accuracy obtained after reducing  the number of kernels in the middle layers to bring down the parameter count to around 8000.
- Run the experiment with the same optimizer as last week, but without rate schedulers.

In [68]:
# load model and summarize 
model1 = models.Model1().to(config.DEVICE)
summary(model1, input_size=(1, 28, 28))

# load the dataloaders
train_loader, test_loader = utils.get_data_loaders()

# initialize the trainer object 
model_trainer = utils.Trainer(model1)
optimizer = optim.Adam(model1.parameters(), lr=config.INIT_LR)
loss_fn = F.nll_loss



----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 28, 28]              72
              ReLU-2            [-1, 8, 28, 28]               0
       BatchNorm2d-3            [-1, 8, 28, 28]              16
           Dropout-4            [-1, 8, 28, 28]               0
            Conv2d-5            [-1, 8, 28, 28]             576
              ReLU-6            [-1, 8, 28, 28]               0
       BatchNorm2d-7            [-1, 8, 28, 28]              16
           Dropout-8            [-1, 8, 28, 28]               0
            Conv2d-9            [-1, 8, 28, 28]             576
             ReLU-10            [-1, 8, 28, 28]               0
      BatchNorm2d-11            [-1, 8, 28, 28]              16
          Dropout-12            [-1, 8, 28, 28]               0
           Conv2d-13            [-1, 8, 28, 28]             576
             ReLU-14            [-1, 8,

<IPython.core.display.Javascript object>

In [69]:
%%time 
model_trainer.run(train_loader=train_loader, test_loader=test_loader, loss_fn=loss_fn, optimizer=optimizer)

1 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:13<00:00, 34.37it/s]

epoch = 1 loss=0.14117105305194855
            train_accuracy=94.64833333333333



Test set: Average loss: 0.0470, Accuracy: 9837/10000 (98%)

2 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.71it/s]

epoch = 2 loss=0.06900804489850998
            train_accuracy=97.51166666666667



Test set: Average loss: 0.0347, Accuracy: 9895/10000 (99%)

3 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:12<00:00, 37.50it/s]

epoch = 3 loss=0.04310191050171852
            train_accuracy=97.93



Test set: Average loss: 0.0312, Accuracy: 9888/10000 (99%)

4 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:12<00:00, 38.84it/s]

epoch = 4 loss=0.06030112877488136
            train_accuracy=98.16833333333334



Test set: Average loss: 0.0300, Accuracy: 9897/10000 (99%)

5 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:13<00:00, 35.85it/s]

epoch = 5 loss=0.038536686450242996
            train_accuracy=98.29166666666667



Test set: Average loss: 0.0286, Accuracy: 9910/10000 (99%)

6 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:13<00:00, 35.02it/s]

epoch = 6 loss=0.08714286237955093
            train_accuracy=98.25166666666667



Test set: Average loss: 0.0256, Accuracy: 9907/10000 (99%)

7 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:13<00:00, 35.15it/s]

epoch = 7 loss=0.02353222481906414
            train_accuracy=98.39333333333333



Test set: Average loss: 0.0415, Accuracy: 9873/10000 (99%)

8 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:13<00:00, 35.99it/s]

epoch = 8 loss=0.060779739171266556
            train_accuracy=98.455



Test set: Average loss: 0.0261, Accuracy: 9915/10000 (99%)

9 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:13<00:00, 34.80it/s]

epoch = 9 loss=0.08816304057836533
            train_accuracy=98.49166666666666



Test set: Average loss: 0.0209, Accuracy: 9934/10000 (99%)

10 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:22<00:00, 21.12it/s]

epoch = 10 loss=0.0639721229672432
            train_accuracy=98.60333333333334



Test set: Average loss: 0.0327, Accuracy: 9902/10000 (99%)

11 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:16<00:00, 28.52it/s]

epoch = 11 loss=0.012499145232141018
            train_accuracy=98.52333333333333



Test set: Average loss: 0.0233, Accuracy: 9930/10000 (99%)

12 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:14<00:00, 32.27it/s]

epoch = 12 loss=0.020866328850388527
            train_accuracy=98.67



Test set: Average loss: 0.0239, Accuracy: 9931/10000 (99%)

13 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:15<00:00, 30.24it/s]

epoch = 13 loss=0.08549397438764572
            train_accuracy=98.64666666666666



Test set: Average loss: 0.0240, Accuracy: 9927/10000 (99%)

14 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:15<00:00, 30.64it/s]

epoch = 14 loss=0.03249802067875862
            train_accuracy=98.79333333333334



Test set: Average loss: 0.0273, Accuracy: 9907/10000 (99%)

15 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:14<00:00, 33.34it/s]

epoch = 15 loss=0.06622859090566635
            train_accuracy=98.76166666666667



Test set: Average loss: 0.0220, Accuracy: 9926/10000 (99%)

CPU times: user 1min 58s, sys: 30.1 s, total: 2min 28s
Wall time: 4min 8s


<IPython.core.display.Javascript object>

## Result

- Total number of parameters = 8,162
- Best Train and Validation accuracy = 98.79% (14th epoch) and 99.34% (9th epoch) respectively

## Analysis

- Naively reducing the number of kernels from the previous submission does not work. I'm repeating a bunch of 3 * 3 Convolution layers after the first layer which stack up the parameter count without increasing the number of kernels available for feature extraction. Experimenting with the number of Maxpool and Conv2d layers might help here. 
- The gap between training and validation accuracy suggests that the model is underfitting. Playing around with the dropout rates might help reduce the gap between the two.
- The validation accuracy does not improve consistently with the number of epochs. It peaks at 99.34% and then drops down again and finally settles at 99.02 in the final epoch. The model performance can be finetuned with varying the optimizer and learning rates, which will be experimented with later.

# Model 2

## Target :- 

- Vary the number of Conv2d / Maxpool layers to bring the parameter count to under 8000, while retaining the model performance of the first model. 
- Reduce model underfitting by playing around with the dropout rates. 

In [73]:
# load model and summarize 
model2 = models.Model2().to(config.DEVICE)
summary(model2, input_size=(1, 28, 28))

# load the dataloaders
train_loader, test_loader = utils.get_data_loaders()

# initialize the trainer object 
model_trainer2 = utils.Trainer(model2)
optimizer = optim.Adam(model2.parameters(), lr=config.INIT_LR)
loss_fn = F.nll_loss



----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 16, 24, 24]           1,152
              ReLU-6           [-1, 16, 24, 24]               0
       BatchNorm2d-7           [-1, 16, 24, 24]              32
         MaxPool2d-8           [-1, 16, 12, 12]               0
           Dropout-9           [-1, 16, 12, 12]               0
           Conv2d-10           [-1, 16, 10, 10]           2,304
             ReLU-11           [-1, 16, 10, 10]               0
      BatchNorm2d-12           [-1, 16, 10, 10]              32
          Dropout-13           [-1, 16, 10, 10]               0
           Conv2d-14             [-1, 1

<IPython.core.display.Javascript object>

In [74]:
%%time 
model_trainer2.run(train_loader=train_loader, test_loader=test_loader, loss_fn=loss_fn, optimizer=optimizer)

1 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:12<00:00, 38.78it/s]

epoch = 1 loss=0.1069692000746727
            train_accuracy=95.79



Test set: Average loss: 0.0437, Accuracy: 9857/10000 (99%)

2 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 43.52it/s]

epoch = 2 loss=0.011444106698036194
            train_accuracy=98.27166666666666



Test set: Average loss: 0.0374, Accuracy: 9875/10000 (99%)

3 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 45.03it/s]

epoch = 3 loss=0.003744215937331319
            train_accuracy=98.51



Test set: Average loss: 0.0321, Accuracy: 9895/10000 (99%)

4 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 40.09it/s]

epoch = 4 loss=0.0012992117553949356
            train_accuracy=98.74833333333333



Test set: Average loss: 0.0286, Accuracy: 9906/10000 (99%)

5 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.56it/s]

epoch = 5 loss=0.0912676677107811
            train_accuracy=98.76666666666667



Test set: Average loss: 0.0399, Accuracy: 9862/10000 (99%)

6 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.44it/s]

epoch = 6 loss=0.01726599968969822
            train_accuracy=98.89666666666666



Test set: Average loss: 0.0307, Accuracy: 9909/10000 (99%)

7 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 44.15it/s]

epoch = 7 loss=0.22413110733032227
            train_accuracy=98.92



Test set: Average loss: 0.0292, Accuracy: 9904/10000 (99%)

8 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 40.45it/s]

epoch = 8 loss=0.04393462836742401
            train_accuracy=98.97833333333334



Test set: Average loss: 0.0296, Accuracy: 9909/10000 (99%)

9 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 43.63it/s]

epoch = 9 loss=0.0018167560920119286
            train_accuracy=98.975



Test set: Average loss: 0.0305, Accuracy: 9911/10000 (99%)

10 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 40.26it/s]

epoch = 10 loss=0.02647802047431469
            train_accuracy=99.02166666666666



Test set: Average loss: 0.0244, Accuracy: 9927/10000 (99%)

11 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.72it/s]

epoch = 11 loss=0.027658618986606598
            train_accuracy=99.125



Test set: Average loss: 0.0235, Accuracy: 9934/10000 (99%)

12 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 41.92it/s]

epoch = 12 loss=0.0020758891478180885
            train_accuracy=99.165



Test set: Average loss: 0.0260, Accuracy: 9912/10000 (99%)

13 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:12<00:00, 38.72it/s]

epoch = 13 loss=0.0442708395421505
            train_accuracy=99.115



Test set: Average loss: 0.0250, Accuracy: 9916/10000 (99%)

14 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 40.74it/s]

epoch = 14 loss=0.0021977717988193035
            train_accuracy=99.16666666666667



Test set: Average loss: 0.0256, Accuracy: 9923/10000 (99%)

15 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:12<00:00, 37.97it/s]

epoch = 15 loss=0.011371631175279617
            train_accuracy=99.14333333333333



Test set: Average loss: 0.0264, Accuracy: 9920/10000 (99%)

CPU times: user 1min 16s, sys: 18.8 s, total: 1min 35s
Wall time: 3min 23s


<IPython.core.display.Javascript object>

## Result

- Total number of parameters = 7,384
- Best Train and Validation accuracy = 99.16% (14th epoch) and 99.34% (11th epoch) respectively

## Analysis

- The maximum validation accuracy, while still lower than the benchmark, is closer to what was achieved in the previous experiment, but with # parameters < 8000. 
- The gap between the training and validation accuracy has narrowed down after experimenting with dropout rates, as compared to the previous model. 
- The validation accuracy jumps around quite a bit during the final epoches, which suggests that the model training might benefit from a lower learning rate during the later epochs. Experimenting with learning rate schedulers should smoothen the training process as well inch closer to hitting the benchmark of 99.4%

# Model 3

## Target 

- Experiment with Learning Rate Schedulers to reduce variation in model validation towards the latter epoches. 
- Hit the validation benchmark of 99.4% consistently during the final training epoches. 

In [78]:
# load model and summarize 
model3 = models.Model3().to(config.DEVICE)
summary(model3, input_size=(1, 28, 28))

# load the dataloaders
train_loader, test_loader = utils.get_data_loaders()

# initialize the trainer object 
model_trainer3 = utils.Trainer(model3)
optimizer = optim.Adam(model3.parameters(), lr=config.INIT_LR)
loss_fn = F.nll_loss


----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 16, 24, 24]           1,152
              ReLU-6           [-1, 16, 24, 24]               0
       BatchNorm2d-7           [-1, 16, 24, 24]              32
         MaxPool2d-8           [-1, 16, 12, 12]               0
           Dropout-9           [-1, 16, 12, 12]               0
           Conv2d-10           [-1, 16, 10, 10]           2,304
             ReLU-11           [-1, 16, 10, 10]               0
      BatchNorm2d-12           [-1, 16, 10, 10]              32
          Dropout-13           [-1, 16, 10, 10]               0
           Conv2d-14             [-1, 1

<IPython.core.display.Javascript object>

In [79]:
%%time 
model_trainer3.run(train_loader=train_loader, test_loader=test_loader, loss_fn=loss_fn, optimizer=optimizer, use_scheduler=True)

1 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 44.64it/s]

epoch = 1 loss=0.025978581979870796
            train_accuracy=95.76666666666667



Test set: Average loss: 0.0413, Accuracy: 9856/10000 (99%)

2 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 40.33it/s]

epoch = 2 loss=0.10401951521635056
            train_accuracy=98.31333333333333



Test set: Average loss: 0.0387, Accuracy: 9871/10000 (99%)

3 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 44.97it/s]

epoch = 3 loss=0.009729928337037563
            train_accuracy=98.72666666666667



Test set: Average loss: 0.0301, Accuracy: 9891/10000 (99%)

4 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 42.96it/s]

epoch = 4 loss=0.037860408425331116
            train_accuracy=98.87



Test set: Average loss: 0.0225, Accuracy: 9922/10000 (99%)

5 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.16it/s]

epoch = 5 loss=0.05278022959828377
            train_accuracy=99.05666666666667



Test set: Average loss: 0.0251, Accuracy: 9922/10000 (99%)

6 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:12<00:00, 38.50it/s]

epoch = 6 loss=0.01982741244137287
            train_accuracy=99.14833333333333



Test set: Average loss: 0.0230, Accuracy: 9935/10000 (99%)

7 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:10<00:00, 43.66it/s]

epoch = 7 loss=0.004092735704034567
            train_accuracy=99.23



Test set: Average loss: 0.0214, Accuracy: 9933/10000 (99%)

8 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.53it/s]

epoch = 8 loss=0.0028035584837198257
            train_accuracy=99.37666666666667



Test set: Average loss: 0.0168, Accuracy: 9949/10000 (99%)

9 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 42.05it/s]

epoch = 9 loss=0.005939218681305647
            train_accuracy=99.40833333333333



Test set: Average loss: 0.0167, Accuracy: 9946/10000 (99%)

10 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.94it/s]

epoch = 10 loss=0.05558159574866295
            train_accuracy=99.46166666666667



Test set: Average loss: 0.0166, Accuracy: 9949/10000 (99%)

11 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 40.76it/s]

epoch = 11 loss=0.009097282774746418
            train_accuracy=99.5



Test set: Average loss: 0.0166, Accuracy: 9945/10000 (99%)

12 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 41.45it/s]

epoch = 12 loss=0.001038345624692738
            train_accuracy=99.55166666666666



Test set: Average loss: 0.0167, Accuracy: 9949/10000 (99%)

13 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 41.47it/s]

epoch = 13 loss=0.01457692589610815
            train_accuracy=99.55166666666666



Test set: Average loss: 0.0168, Accuracy: 9949/10000 (99%)

14 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 42.04it/s]

epoch = 14 loss=0.002278827363625169
            train_accuracy=99.55833333333334



Test set: Average loss: 0.0168, Accuracy: 9944/10000 (99%)

15 / 15


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 469/469 [00:11<00:00, 39.66it/s]

epoch = 15 loss=0.005205374211072922
            train_accuracy=99.52666666666667



Test set: Average loss: 0.0169, Accuracy: 9950/10000 (100%)

CPU times: user 1min 17s, sys: 19.2 s, total: 1min 37s
Wall time: 3min 22s


<IPython.core.display.Javascript object>

## Result

- Total number of parameters = 7,384
- Best Train and Validation accuracy = 99.56% (14th epoch) and 99.5% (15th epoch) respectively

## Analysis

- The model hits the benchmark in the 8th epoch, and remains above it till the training ends, indicating the model might improve further with additional training.
- The gap between training and validation accuracy improves over the previous model.
